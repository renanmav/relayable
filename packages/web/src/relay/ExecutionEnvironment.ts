const canUseDOM = !!(
  typeof window !== 'undefined' &&
  window.document &&
  window.document.createElement
)

const ExecutionEnvironment = {
  canUseDOM,
  canUseWorkers: typeof Worker !== 'undefined',
  canUseEventListeners:
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    canUseDOM && !!(window.addEventListener || (window as any).attachEvent),
  canUseViewport: canUseDOM && !!window.screen,
  isInWorker: !canUseDOM,
}

export default ExecutionEnvironment
