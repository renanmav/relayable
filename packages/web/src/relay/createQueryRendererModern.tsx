import React from 'react'
import { GraphQLTaggedNode, Variables } from 'relay-runtime'
import { QueryRenderer } from 'react-relay'
import hoistStatics from 'hoist-non-react-statics'
// @ts-ignore
import { createMockEnvironment } from 'relay-test-utils'

import { Environment } from '.'

/* eslint-disable @typescript-eslint/no-explicit-any */
interface Config {
  query?: GraphQLTaggedNode
  queriesParams?: (props: Object) => Object
  variables?: Variables
  hideSplash?: boolean
}

export const env = createMockEnvironment()

export default function createQueryRenderer(
  FragmentComponent: React.ComponentType<any>,
  Component: React.ComponentType<any>,
  config: Config,
  LoadingComponent?: React.ComponentType<any>
) {
  const { query, queriesParams } = config

  function QueryRendererWrapper(props: Object) {
    const variables = (queriesParams ? queriesParams(props) : config.variables) as Variables

    return (
      <QueryRenderer
        environment={process.env.NODE_ENV === 'test' ? env : Environment}
        query={query}
        variables={variables}
        render={({ error, props: relayProps }) => {
          if (error) {
            return (
              <>
                <h1>{error.name.toString()}</h1>
                <p>{error.message.toString()}</p>
              </>
            )
          }

          if (relayProps) {
            return <FragmentComponent {...props} query={relayProps} />
          }

          return LoadingComponent ? <LoadingComponent /> : <div>Loading</div>
        }}
      />
    )
  }

  return hoistStatics(QueryRendererWrapper, Component)
}
