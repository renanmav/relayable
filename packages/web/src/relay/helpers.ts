/* eslint-disable @typescript-eslint/no-explicit-any */
import {
  RequestParameters,
  Variables,
  UploadableMap,
  CacheConfig
} from 'relay-runtime'

export const isMutation = (request: RequestParameters) => request.operationKind === 'mutation'
export const isQuery = (request: RequestParameters) => request.operationKind === 'query'
export const forceFetch = (cacheConfig: CacheConfig) => !!(cacheConfig && cacheConfig.force)

function getRequestBodyWithUploadables (
  request: RequestParameters,
  variables: Variables,
  uploadables: UploadableMap
) {
  const formData = new FormData()
  formData.append('name', request.name)
  // @ts-ignore
  formData.append('query', request.text)
  formData.append('variables', JSON.stringify(variables))

  Object.keys(uploadables).forEach(key => {
    if (Object.prototype.hasOwnProperty.call(uploadables, key)) {
      formData.append(key, uploadables[key])
    }
  })

  return formData
}

function getRequestBodyWithoutUploadables (
  request: RequestParameters,
  variables: Variables
) {
  return JSON.stringify({
    name: request.name,
    query: request.text,
    variables
  })
}

export function getRequestBody (
  request: RequestParameters,
  variables: Variables,
  uploadables?: UploadableMap
) {
  if (uploadables) {
    return getRequestBodyWithUploadables(request, variables, uploadables)
  }

  return getRequestBodyWithoutUploadables(request, variables)
}

export const getHeaders = (uploadables?: UploadableMap) => {
  if (uploadables) {
    return {
      Accept: '*/*'
    }
  }

  return {
    Accept: 'application/json',
    'Content-type': 'application/json'
  }
}

export const handleData = (response: any) => {
  const contentType = response.headers.get('content-type')
  if (contentType && contentType.indexOf('application/json') !== -1) {
    return response.json()
  }

  return response.text()
}
