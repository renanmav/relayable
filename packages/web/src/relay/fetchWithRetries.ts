/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-use-before-define */
import ExecutionEnvironment from './ExecutionEnvironment'

export type InitWithRetries = {
  body?: unknown
  cache?: string | null
  credentials?: string | null
  fetchTimeout?: number | null
  headers?: unknown
  method?: string | null
  mode?: string | null
  retryDelays?: Array<number> | null
}

const DEFAULT_TIMEOUT = 15000
const DEFAULT_RETRIES = [1000, 3000]

function fetchWithRetries (
  uri: string,
  initWithRetries?: InitWithRetries | null
): Promise<any> {
  // @ts-ignore
  const { fetchTimeout, retryDelays, ...init } = initWithRetries || {}
  const _fetchTimeout = fetchTimeout !== null ? fetchTimeout : DEFAULT_TIMEOUT
  const _retryDelays = retryDelays !== null ? retryDelays : DEFAULT_RETRIES

  let requestsAttempted = 0
  let requestStartTime = 0
  return new Promise((resolve, reject) => {
    function shouldRetry (attempt: number): boolean {
      return ExecutionEnvironment.canUseDOM && attempt <= _retryDelays!.length
    }

    function sendTimedRequest (): void {
      requestsAttempted++
      requestStartTime = Date.now()
      let isRequestAlive = true
      // @ts-ignore
      const request = fetch(uri, init)
      const requestTimeout = setTimeout(() => {
        isRequestAlive = false
        if (shouldRetry(requestsAttempted)) {
          // eslint-disable-next-line no-console
          console.log(false, 'fetchWithRetries: HTTP timeout, retrying.')
          retryRequest()
        } else {
          reject(
            new Error(
              `fetchWithRetries(): Failed to get response from server, tried ${requestsAttempted} times.`
            )
          )
        }
      }, _fetchTimeout)

      request
        // @ts-ignore
        .then(response => {
          clearTimeout(requestTimeout)
          if (isRequestAlive) {
            if (response.status >= 200 && response.status < 300) {
              resolve(response)
            } else if (response.status === 401) {
              resolve(response)
            } else if (shouldRetry(requestsAttempted)) {
              // eslint-disable-next-line no-console
              console.log(false, 'fetchWithRetries: HTTP error, retrying.')
              retryRequest()
            } else {
              // Request was not successful, giving up.
              const error: any = new Error(
                `fetchWithRetries(): Still no successful response after ${requestsAttempted} retries, giving up.`
              )
              error.response = response
              reject(error)
            }
          }
        })
        // @ts-ignore
        .catch(error => {
          clearTimeout(requestTimeout)
          if (shouldRetry(requestsAttempted)) {
            retryRequest()
          } else {
            reject(error)
          }
        })
    }

    function retryRequest (): void {
      const retryDelay = _retryDelays![requestsAttempted - 1]
      const retryStartTime = requestStartTime + retryDelay
      setTimeout(sendTimedRequest, retryStartTime - Date.now())
    }

    sendTimedRequest()
  })
}

export default fetchWithRetries
