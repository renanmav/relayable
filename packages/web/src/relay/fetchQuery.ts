import { RequestParameters, Variables, UploadableMap } from 'relay-runtime'

import { yottaToken } from '../utils/contants'

import { getRequestBody, getHeaders, handleData, isMutation } from './helpers'
import fetchWithRetries from './fetchWithRetries'

// eslint-disable-next-line prefer-destructuring
export const GRAPHQL_URL = process.env.GRAPHQL_URL || 'http://localhost:5000/graphql'

const fetchQuery = async (
  request: RequestParameters,
  variables: Variables,
  uploadables: UploadableMap
) => {
  try {
    const body = getRequestBody(request, variables, uploadables)
    const headers = {
      ...getHeaders(uploadables),
      authorization: localStorage.getItem(yottaToken),
    }

    const response = await fetchWithRetries(GRAPHQL_URL, {
      method: 'POST',
      headers,
      body,
      fetchTimeout: 20000,
      retryDelays: [1000, 3000, 5000],
    })

    const data = await handleData(response)

    if (response.status === 401) {
      throw data.errors
    }

    if (isMutation(request) && data.errors) {
      throw data
    }

    if (!data.data) {
      throw data.errors
    }

    return data
  } catch (err) {
    // eslint-disable-next-line no-console
    console.log('err: ', err)

    const timeoutRegexp = new RegExp(/Still no successful response after/)
    const serverUnavailableRegexp = new RegExp(/Failed to fetch/)
    if (timeoutRegexp.test(err.message) || serverUnavailableRegexp.test(err.message)) {
      throw new Error('Serviço indisponível. Tente novamente mais tarde.')
    }

    throw err
  }
}

export default fetchQuery
