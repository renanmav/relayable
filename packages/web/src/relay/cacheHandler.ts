import RelayQueryResponseCache from 'relay-runtime/lib/network/RelayQueryResponseCache'
import { FetchFunction } from 'relay-runtime'

import { isMutation, isQuery, forceFetch } from './helpers'
import fetchQuery from './fetchQuery'

const oneMinute = 60 * 1000
const relayResponseCache = new RelayQueryResponseCache({
  size: 250,
  ttl: oneMinute,
})

const cacheHandler: FetchFunction = async (request, variables, cacheConfig, uploadables) => {
  const queryID = request.text

  if (isMutation(request)) {
    relayResponseCache.clear()
    return fetchQuery(request, variables, uploadables!)
  }

  const fromCache = relayResponseCache.get(queryID!, variables)
  if (isQuery(request) && fromCache !== null && !forceFetch(cacheConfig)) {
    return fromCache
  }

  const fromServer = await fetchQuery(request, variables, uploadables!)
  if (fromServer) {
    relayResponseCache.set(queryID!, variables, fromServer)
  }

  return fromServer
}

export default cacheHandler
