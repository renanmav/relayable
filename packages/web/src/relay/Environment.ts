import { Environment, Network, RecordSource, Store } from 'relay-runtime'
// @ts-ignore
import RelayNetworkLogger from 'relay-runtime/lib/RelayNetworkLogger'
// @ts-ignore
import { installRelayDevTools } from 'relay-devtools'

import cacheHandler from './cacheHandler'

if (process.env.NODE_ENV === 'development') {
  installRelayDevTools()
}

const network = Network.create(
  process.env.NODE_ENV === 'development' ? RelayNetworkLogger.wrapFetch(cacheHandler) : cacheHandler
)

const source = new RecordSource()
const store = new Store(source)

const env = new Environment({
  network,
  store,
})

export default env
