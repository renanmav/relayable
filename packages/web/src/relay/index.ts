import Environment from './Environment'
import createQueryRendererModern from './createQueryRendererModern'

export { Environment, createQueryRendererModern }
