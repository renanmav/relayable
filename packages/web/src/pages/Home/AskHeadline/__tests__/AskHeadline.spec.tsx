import React from 'react'
import renderer from 'react-test-renderer'
import { env } from '@yotta/web/src/relay/createQueryRendererModern'
// @ts-ignore
import { MockPayloadGenerator } from 'relay-test-utils'

// @ts-ignore
import { RelayEnvironmentProvider } from '@entria/relay-experimental'

import Ask from '../index'

const Providers: React.FC = ({ children }) => (
  <RelayEnvironmentProvider environment={env}>
    <React.Suspense fallback={<div>Loading...</div>}>{children}</React.Suspense>
  </RelayEnvironmentProvider>
)

test('Loading state', () => {
  const component = renderer.create(
    <Providers>
      <Ask />
    </Providers>
  )

  expect(component).toMatchSnapshot()
})

test('Data Render', () => {
  const component = renderer.create(
    <Providers>
      <Ask />
    </Providers>
  )

  env.mock.resolveMostRecentOperation((operation: any) => MockPayloadGenerator.generate(operation))

  expect(component).toMatchSnapshot()
})
