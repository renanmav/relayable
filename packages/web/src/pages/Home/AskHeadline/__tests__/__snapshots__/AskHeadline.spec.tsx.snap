// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Render 1`] = `
<div>
  Loading...
</div>
`;

exports[`Loading state 1`] = `
<div>
  Loading...
</div>
`;
